### ISSUE TITLE
Duplicated sample points returned by query-frontend

### ISSUE THREAD

**Issue by alvinlin123** [cortexproject/cortex#3920]
**Describe the bug**

I have continuous test running that queries Cortex. After upgrading from commit https://github.com/cortexproject/cortex/commit/af9e20c54ee97f409008f3e86541c5dfa5038e22 to  https://github.com/cortexproject/cortex/commit/dd6dbf9c03a13c6a3322a4a05b2261b0c13f5107 we observed that around 16:00:00 pacific time our test would fail because duplicate sample points were returned: 

```
curl --location -g --request POST 'http://localhost:8080/prometheus/api/v1/query_range?query=my_metrics_name&start=1614556800&end=1614557160&step=60s'
{
  "status": "success",
  "data": {
    "resultType": "matrix",
    "result": [
      {
        "metric": {
          "__name__": "my_metrics_name"
        },
        "values": [
          [
            1614556800,
            "96"
          ],
          [
            1614556860,
            "141"
          ],
          [
            1614556800,
            "96"
          ],
          [
            1614556860,
            "141"
          ],
          [
            1614556920,
            "141"
          ],
          [
            1614556980,
            "141"
          ],
          [
            1614557040,
            "141"
          ],
          [
            1614557100,
            "141"
          ],
          [
            1614557160,
            "185"
          ]
        ]
      }
    ]
  }
} 
```
Notice how sample point with timestamp `1614556800` and `1614556860` are duplicated. 

Our continuous is quite simple, it runs periodically to push some data and query some data from Cortex. Our test is not using HA mode.  

**To Reproduce**
I am not sure how to reproduce it, but one  pattern is that it seems to happen everyday around 16:00:00 pacific time (00:00:00 GMT). The query I had issues with are queries that query past data points like:
* query data from 24 hours ago to 24 hours + 6 minutes ago (this query will land in ingester). 
* query data from 48 hours ago to 48 hours + 6 minuges ago (this query will not land in ingester).

Also, if I keep making the same query multiple times, the issues goes away after a few tries for that specific query; problem comes back next day at 16:00:00 pacific time. 

**Expected behavior**
I don't expected sample points to be duplicated because: 
1. Cortex rejects duplicated sample points for write path when not using HA mode.
2. Query path should deduplicate result from different sources.

**Environment:**
 - Infrastructure: k8s
 - Deployment tool: helm
- I am using query front-end  

**Storage Engine**
- [X] Blocks
- [ ] Chunks

**Comment by bboreham** [comment]
Could you clarify whether you are using the `query-frontend` component, and if so, whether the symptom persists if you call the `querier` directly?

**Comment by jeromeinsf** [comment]
Yes query-frontend is being used.

**Comment by alvinlin123** [comment]
> Could you clarify whether you are using the `query-frontend` component, and if so, whether the symptom persists if you call the `querier` directly?

I tried query all the queries directly after I see the bug through `query-frontend`, but was not able to reproduce the bug.

**Comment by alvinlin123** [comment]
One more data point. I was able to reproduce the issue by querying agains `query-frontend` only as soon as I see duplicate sample points querying `query-frontend`, I made the exact same query against all the `queriers`, but `querier` never return duplicated samples. So basically here is what I did: 

1. port-forward to `query-frontend-x`
2. make `query_range` against the `query-frontend`, got duplicate samples
3. make the same `query_range` call against ALL The querier, didn't get duplicate samples
4. make the same `query_range` call agains the same `query-frontend`, got duplicate samples again.

**Comment by jeromeinsf** [comment]
From the Cortex community call, wondering if this would be in or around https://github.com/cortexproject/cortex/blob/master/pkg/querier/queryrange/split_by_interval.go#L43

**Comment by alvinlin123** [comment]
ok, one more datapoint I just discovered, if I set `cache_results` to true I get the duplicate samples, if I set `cache_results` to false, I can't re-produce the duplicate sample issue hmmm.

**Comment by alvinlin123** [comment]
I just noticed something weird. With old commit https://github.com/cortexproject/cortex/commit/af9e20c54ee97f409008f3e86541c5dfa5038e22 the size of cache entry doesn't change. But after I upgrade to commit https://github.com/cortexproject/cortex/commit/dd6dbf9c03a13c6a3322a4a05b2261b0c13f5107, I started to see cache item size start to grow everytime I make a same query

```
key=cc4fd9127b36005e exp=-1 la=1615952893 cas=1181 fetch=no cls=21 size=7806
key=cc4fdc127b360577 exp=-1 la=1615952893 cas=1182 fetch=no cls=22 size=10308
END
lru_crawler metadump all
key=cc4fd9127b36005e exp=-1 la=1615952896 cas=1183 fetch=no cls=21 size=7860
key=cc4fdc127b360577 exp=-1 la=1615952896 cas=1184 fetch=no cls=22 size=10380
END
lru_crawler metadump all
key=cc4fd9127b36005e exp=-1 la=1615952896 cas=1183 fetch=no cls=21 size=7860
key=cc4fdc127b360577 exp=-1 la=1615952896 cas=1184 fetch=no cls=22 size=10380
END
lru_crawler metadump all
key=cc4fd9127b36005e exp=-1 la=1615952902 cas=1188 fetch=no cls=21 size=7968
key=cc4fdc127b360577 exp=-1 la=1615952902 cas=1187 fetch=no cls=22 size=10524
END
lru_crawler metadump all
key=cc4fd9127b36005e exp=-1 la=1615952902 cas=1188 fetch=no cls=21 size=7968
key=cc4fdc127b360577 exp=-1 la=1615952902 cas=1187 fetch=no cls=22 size=10524
END
lru_crawler metadump all
key=cc4fd9127b36005e exp=-1 la=1615952908 cas=1191 fetch=no cls=21 size=8076
key=cc4fdc127b360577 exp=-1 la=1615952908 cas=1192 fetch=no cls=22 size=10668
END
lru_crawler metadump all
key=cc4fd9127b36005e exp=-1 la=1615952908 cas=1191 fetch=no cls=21 size=8076
key=cc4fdc127b360577 exp=-1 la=1615952908 cas=1192 fetch=no cls=22 size=10668
END
lru_crawler metadump all
key=cc4fd9127b36005e exp=-1 la=1615952911 cas=1193 fetch=no cls=21 size=8130
key=cc4fdc127b360577 exp=-1 la=1615952911 cas=1194 fetch=no cls=22 size=10740
END
```
note that when `cas` changes,`size` increases. I don't think this is normal, is it? I am suspecting this commit https://github.com/cortexproject/cortex/commit/50ab740d9164aa494d8d09efcc1a08cc0d280f8d result in increase cache entry size, but I need to take a closer look.

With the above in mind, I was somehow able to reproduce the issue if I down grade from newer commit to older commit; and I guess that is because the cached entry contains lots of duplicate samples?

**Comment by pracucci** [comment]
> I am suspecting this commit 50ab740 introduces increase cache entry size behaviour, but I need to take a closer look.

We noticed another issue with the behaviour introduced by 50ab740 and Goutham fixed it in https://github.com/cortexproject/cortex/pull/3818. Could you test if #3818 solves your issue too? I just would like to exclude the issue has already be (unintentionally) fixed in `master`.

**Comment by alvinlin123** [comment]
I actually tested with https://github.com/cortexproject/cortex/pull/3818 in. Specifically, I was testing with https://github.com/cortexproject/cortex/commit/e02797ac7f3b68f08c7b778c95c60ba82303f81b and see the cache entry size increase.

**Comment by alvinlin123** [comment]
Ok, so after looking closer indeed https://github.com/cortexproject/cortex/commit/50ab740d9164aa494d8d09efcc1a08cc0d280f8d caused the issue where cache entry size kept on increasing. Consider the following: 

We are making a range query where 
* start: 1615161360000
* end: 1615161720000
* step: 60000 (1 minute)

Because the above timestamp crosses day's boundary, so the query splitter will split it to two queries:
* Start: 1615161360000, End:   1615161540000 (3 minutes interval)
* Start: 1615161600000, End:   1615161720000 (2 minutes interval)

Note that all the queries has interval less than 5 minutes. So these 2 queries will get pass further down to the `result_cache`  middleware. Let's assume we have made the same query before, so there are  result cache entries for the 2 queries.

Both query will go through the [`handleHit`](https://github.com/cortexproject/cortex/blob/b0424b9e0d6d31f64e71e14ff853f3013f283c6e/pkg/querier/queryrange/results_cache.go#L353) method with existing cache data, then go through the [`partition`](https://github.com/cortexproject/cortex/blob/b0424b9e0d6d31f64e71e14ff853f3013f283c6e/pkg/querier/queryrange/results_cache.go#L477) method . 

Because the queries are less than 5 minutes [this line of code](https://github.com/cortexproject/cortex/blob/b0424b9e0d6d31f64e71e14ff853f3013f283c6e/pkg/querier/queryrange/results_cache.go#L493) will not use the cache data, and will return an `Request` object for each query. 

`handleHit` then go ahead and execute the `Request` returned by `partition` and append it to the existing cached data. But now the problem comes; existing cached data already have the exactly same data, by appending the result of the `Request` we created a duplicated `Extent` in the cache entry. The more we execute the same query, the more duplicate `Extent` gets added to the cache entry; hence, the ever growing cache entry size.

Before https://github.com/cortexproject/cortex/commit/50ab740d9164aa494d8d09efcc1a08cc0d280f8d, I think it was expected that `partition` method returns requests to get data that is **not** in the cache. Commit https://github.com/cortexproject/cortex/commit/50ab740d9164aa494d8d09efcc1a08cc0d280f8d breaks that expectation because after the change, `partition` would return requests to get data that is in cache. 

I am working on a PR to fix this.

**Comment by bboreham** [comment]
What is left to fix here after #3968?

**Comment by alvinlin123** [comment]
I am not seeing anymore duplicate samples pointing now, so I will go ahead and resolve this issue now.
